src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -5387,6 +5387,76 @@ static QVariant fcnAzimuth( const QVariantList &values, const QgsExpressionConte
   }
 }
 
+static QVariant fcnBearing( const QVariantList &values, const QgsExpressionContext *context, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  const QgsGeometry geom1 = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  const QgsGeometry geom2 = QgsExpressionUtils::getGeometry( values.at( 1 ), parent );
+  const QString sAuthId = QgsExpressionUtils::getStringValue( values.at( 2 ), parent );
+  const QString ellipsoid = QgsExpressionUtils::getStringValue( values.at( 3 ), parent );
```

Comment:
The ellipsoid and source crs could be made optional (or removed?) if the distance area calculator were obtained using `QgsDistanceArea* calc = parent->geomCalculator()`. (That's what other functions like area/distance/perimeter use)

Reply:
Thanks for the suggestion.
geomCalculator() - which I've initially considered to use -  is used in the "$" functions which don't have parameters and directly use the current feature geometry from the layer as input, so it is necessary and logic to infer the CRS and the ellipsoid from the layer and from the project.
bearing function, like e.g. transform function, can have input points not necessary in the CRS of the layer (which could be a non geospatial layer containing the coordinates of the points in fields values) so I would maintain the possibility to specify the CRS and also the ellipsoid (to give full control to the user).
I would try to make the source_auth_id and the ellipsoid parameters optional: if source_auth_id is not given, then the layer CRS is used, and if ellipsoid is not given, then the project ellipsoid is used.

Reply:
- source_auth_id and the ellipsoid parameters made optional and function help updated accordingly

- code cleaned up:
no need of 
```
  if ( geom1.isMultipart() )
  {
    const QgsMultiPointXY multiPoint = geom1.asMultiPoint();
    if ( multiPoint.count() == 1 )
    {
      point1 = multiPoint[0];
    }
  }
```
since QgsGeometry.asPoint() already does the job with the same results

- avoided to return non finite value (since geod_inverse used in QgsDistanceArea::bearing can return a NaN value in some circumstances)
- additional tests added